### PR TITLE
feat: Add per-tool enable/disable toggle for MCP tools

### PR DESCRIPTION
**Pull request by hatayama**
## Summary

- Add UI toggle in Unity Editor (Window > uLoop > Tool Settings) to enable/disable individual MCP tools
- Disabled tools are completely hidden from AI agents: removed from MCP tool list, CLI commands, and skill files
- Settings stored in `.uloop/settings.tools.json` (project-scoped, blacklist approach)
- Built-in and third-party tools are displayed in separate groups
- Internal tools (`DisplayDevelopmentOnly=true`) are excluded from the toggle UI and always enabled

## Related Issues

- Related to #635 (project-scoped settings): Tool settings use the same `.uloop/` directory pattern for project-local configuration

## Key Implementation Details

- **Data layer**: `ToolSettings.cs` with cached settings, atomic writes, and `ToolSettingsData` record
- **Tool filtering**: `UnityToolRegistry.GetRegisteredTools()` filters disabled tools; `ExecuteToolAsync()` throws `ToolDisabledException` with user-friendly message pointing to the settings UI
- **CLI filtering**: `tool-settings-loader.ts` provides `filterEnabledTools()` with `--project-path` support; all command registration/lookup in `cli.ts` respects disabled tools
- **Skill sync**: `ToolSkillSynchronizer.cs` removes skill directories on disable and reinstalls on enable; SKILL.md `toolName` frontmatter field maps tool names to skill names (handles aliases like `capture-window` → `uloop-screenshot`)
- **UI**: `ToolSettingsSection.cs` with foldout, grouped toggles, deferred heavy work for instant feedback
- **MCP notification**: `tools/list_changed` sent on toggle change

## Tests

- C# unit tests: `ToolSettingsTests.cs` (8 tests), `DefaultErrorTranslatorTests.cs` (7 tests)
- TypeScript unit tests: `tool-settings-loader.test.ts` (18 tests including projectPath override)

## Test Plan

- [ ] Open Window > uLoop > Tool Settings and verify built-in/third-party grouping
- [ ] Toggle a tool OFF → verify `.uloop/settings.tools.json` updated
- [ ] Reconnect MCP client → verify disabled tool is not listed
- [ ] Run `uloop <disabled-tool>` → verify clear disabled-tool error message
- [ ] Run `uloop -h` → verify disabled tool not shown
- [ ] Toggle tool ON → verify skill files reinstalled and tool reappears after MCP reconnect
- [ ] Run `uloop skills install --claude` → verify disabled tool's skill is not installed

## Overview

This PR implements per-tool enable/disable toggles for MCP tools. A new Unity Editor UI (Window > uLoop > Tool Settings) lets users enable/disable individual tools. Disabled tools are excluded from MCP tool lists, CLI command discovery/help, and skill installation; settings are persisted per-project in .uloop/settings.tools.json using a blacklist of disabled tool names.

### Settings Storage & Data Layer
- ToolSettingsData (serializable record) with disabledTools:string[].
- ToolSettings (static class): lazy-loading cache, atomic writes with .bak recovery, size checks, backup-based recovery, public API:
  - GetSettings(), SaveSettings(), IsToolEnabled(), SetToolEnabled(), GetDisabledTools(), InvalidateCache().

### Tool Registry & Execution
- UnityToolRegistry:
  - GetRegisteredTools() filters out disabled tools.
  - GetAllRegisteredToolInfos() returns all tools (for UI).
  - IsThirdPartyTool(toolName) determines third-party vs built-in.
  - Execution path throws ToolDisabledException early if tool is disabled.

### CLI & TypeScript Layer
- tool-settings-loader.ts:
  - loadDisabledTools(projectPath?), isToolEnabled(toolName, projectPath?), filterEnabledTools(tools, projectPath?).
  - Robust parsing: handles missing/empty/invalid JSON, CRLF frontmatter cases, and returns [] on errors.
  - Supports --project-path to scope settings per project.
- cli.ts:
  - Threads projectPath through completion/option listing and command discovery.
  - Registers only enabled tools; unknown-command handling checks enabled set and prints a user-facing disabled-tool error when appropriate.
  - Avoids unnecessary project scanning when projectPath is not provided.

### Skill Synchronization
- ToolSkillSynchronizer (editor utility):
  - RemoveSkillFiles(toolName): removes skill dirs matching tool via SKILL.md frontmatter toolName or fallback naming (uloop-{toolName}).
  - InstallSkillFiles(): invokes skills install per detected target, drains process streams, logs non-zero exit codes, runs asynchronously.
  - Robust frontmatter parsing including CRLF handling.
- skills-manager.ts:
  - SkillDefinition.toolName? populated from frontmatter.
  - isSkillDisabledByToolSettings(skill, disabledTools) used to skip installs/uninstalls for disabled tools.
  - Adds deprecated-skill cleanup support for internal/removed skills.

### Error Handling
- ToolDisabledException class with ToolName property.
- ImprovedErrorHandler maps ToolDisabledException to a user-friendly message and Medium severity; DefaultErrorTranslator tests updated for messaging and menu guidance.

### Editor UI
- View data: ToolToggleItem and ToolSettingsSectionData records.
- ToolSettingsSection UI component: foldout, grouped Built-in/Third-party toggles, UpdateSingleToggle for fast feedback, events OnFoldoutChanged and OnToolToggled.
- McpEditorWindow / McpEditorWindowUI:
  - Integrates ToolSettingsSection, wires events, constructs ToolSettingsSectionData by enumerating registry (excluding development-only/internal tools via DisplayDevelopmentOnly=true).
  - ApplyToolToggle updates ToolSettings, emits tools/list_changed, and defers heavy work (skill install/remove) to avoid UI lag.
- McpEditorModel and McpEditorWindowState updated to persist/show tool settings UI state.
- UXML/USS templates added for the Tool Settings section (note: some style block duplication present).

### Notifications & Server
- TypeScriptServer: tools/list_changed notification behavior updated (removed single-send lifetime gate); 'claude' added to LIST_CHANGED_SUPPORTED_CLIENTS.
- MCP notification emitted when toggles change.

### Supporting Changes & Cleanups
- McpConstants: ULOOP_TOOL_SETTINGS_FILE_NAME = "settings.tools.json".
- McpUIConstants: TOOL_SETTINGS_MENU_PATH = "Window > uLoop > Tool Settings".
- McpEditorSettings: showToolSettings flag with getters/setters.
- Removed public tool entry get-provider-details from default-tools.json and marked its skill/internal status:
  - Added deprecated skill entry uloop-get-provider-details for cleanup.
  - UnitySearchProviderDetailsTool marked DisplayDevelopmentOnly=true so it is excluded from UI toggles and always enabled.
- Tests added/updated to cover behavior and edge cases (see Testing).

## Tests & Reliability Fixes

- C# tests:
  - ToolSettingsTests.cs: tests for enable/disable toggles, persistence, deduplication, cache invalidation reload, backup recovery, multi-tool independence; improved test isolation to clean leftover files.
  - DefaultErrorTranslatorTests.cs: tests for ToolDisabledException translation, messages, and severity mapping.
- TypeScript tests:
  - tool-settings-loader.test.ts: tests for loadDisabledTools (valid/missing/invalid/empty), isToolEnabled, filterEnabledTools, and projectPath override behavior; isolates project-root via temporary dirs.
- Commit fixes:
  - CRLF handling in frontmatter parsing.
  - Guard null disabledTools after JSON deserialization.
  - Improved ToolSkillSynchronizer.RunSkillsInstall to properly drain redirected streams, use using patterns, and log non-zero exit codes.
  - Test cleanup improvements to remove leftover files and ensure isolation.

## Special Cases & Notes

- Internal/development-only tools (DisplayDevelopmentOnly=true) are excluded from the Tool Settings UI but remain enabled and unaffected by user toggles.
- Disabling a tool removes associated skill directories (including aliased mappings via toolName frontmatter) and emits tools/list_changed; enabling reinstalls skills.
- Settings are project-scoped via .uloop/settings.tools.json; CLI supports --project-path to respect those settings.
- Some UI style rules were added in multiple places resulting in duplicated blocks (should be harmless but noted).

## Files of Interest (high level)
- New/modified core: Packages/src/Editor/Config/ToolSettings.cs, ToolSettingsData.cs, ToolSkillSynchronizer.cs, ToolDisabledException.cs, UnityToolRegistry.cs
- UI: Packages/src/Editor/UI/... (ToolSettingsSection.cs, McpEditorWindow*, view data, UXML/USS)
- CLI/TS: Packages/src/Cli~/src/tool-settings-loader.ts, cli.ts, skills-manager.ts, tool-tests
- Tests: Assets/Tests/Editor/ToolSettingsTests.cs, DefaultErrorTranslatorTests.cs; Packages/src/Cli~/src/__tests__/tool-settings-loader.test.ts

### CODE REVIEW

File: Packages/src/Cli~/src/tool-settings-loader.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+export function loadDisabledTools(projectPath?: string): string[] {
+  const projectRoot: string | null =
+    projectPath !== undefined ? resolve(projectPath) : findUnityProjectRoot();
+  if (projectRoot === null) {
+    return [];
+  }
+
+  const settingsPath: string = join(projectRoot, ULOOP_DIR, TOOL_SETTINGS_FILE);
+  if (!existsSync(settingsPath)) {
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Packages/src/Cli~/src/tool-settings-loader.ts]: Fixed in 273bebe. Removed existsSync check and wrapped readFileSync in try-catch to eliminate the TOCTOU race. The file-missing case now falls through to the catch block and returns [] gracefully, consistent with the existing JSON.parse try-catch in the same function.

File: Packages/src/Editor/Config/ToolSettings.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                File.Move(backupPath, SettingsFilePath);
+            }
+
+            if (File.Exists(SettingsFilePath))
+            {
+                FileInfo fileInfo = new FileInfo(SettingsFilePath);
+                Debug.Assert(fileInfo.Length <= McpConstants.MAX_SETTINGS_SIZE_BYTES,
+                    $"Settings file exceeds size limit: {fileInfo.Length} bytes");
+
+                string json = File.ReadAllText(SettingsFilePath);
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Packages/src/Editor/Config/ToolSettings.cs:108]: Not adopting this suggestion. This project follows a strict Fail Fast policy where try-catch is prohibited by default. ToolSettings.cs intentionally uses the same pattern as ULoopSettings.cs (File.Exists → File.ReadAllText → JsonUtility.FromJson without try-catch), which was already reviewed and merged. Corrupted settings in the Unity Editor context should surface immediately via assert/exception rather than being silently swallowed.

File: Packages/src/Cli~/src/cli.ts [hunk 3]
```diff
@@ -822,7 +830,7 @@ async function main(): Promise<void> {
 
   if (skipProjectDetection) {
     const defaultTools = getDefaultTools();
-    for (const tool of defaultTools.tools) {
+    for (const tool of filterEnabledTools(defaultTools.tools, syncGlobalOptions.projectPath)) {
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Packages/src/Cli~/src/cli.ts]: Fixed in 273bebe. Tool filtering in the skipProjectDetection path (and completion handlers) now only applies when --project-path is explicitly provided. Without it, all tools are registered without triggering findUnityProjectRoot() scanning.

File: Assets/Tests/Editor/ToolSettingsTests.cs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+            _backupFileContent = _backupFileExisted ? File.ReadAllText(SettingsBackupPath) : null;
+
+            string uloopDir = Path.GetDirectoryName(SettingsFilePath);
+            if (!string.IsNullOrEmpty(uloopDir) && !Directory.Exists(uloopDir))
+            {
+                Directory.CreateDirectory(uloopDir);
+            }
+
+            ToolSettings.InvalidateCache();
+        }
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Assets/Tests/Editor/ToolSettingsTests.cs:37]: Fixed in 65f6683. Added DeleteIfExists for both primary and backup files in SetUp before InvalidateCache, ensuring tests start with a clean slate regardless of pre-existing state.

File: Packages/src/Cli~/src/skills/skills-manager.ts [hunk 5]
```diff
@@ -239,10 +241,13 @@ function scanEditorFolderForSkills(
         }
 
         const name = typeof frontmatter.name === 'string' ? frontmatter.name : entry.name;
+        const toolName =
+          typeof frontmatter.toolName === 'string' ? frontmatter.toolName : undefined;
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Packages/src/Cli~/src/skills/skills-manager.ts:245]: Fixed in 65f6683. Updated parseFrontmatter regex to use \r?\n for delimiters and split lines with /\r?\n/, matching the CRLF fix already applied to the C# side (ToolSkillSynchronizer) in f5b1711.

File: Packages/src/Editor/Api/McpTools/CaptureWindow/Skill/SKILL.md [hunk 6]
```diff
@@ -1,5 +1,6 @@
 ---
 name: uloop-screenshot
+toolName: capture-window
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Packages/src/Editor/Api/McpTools/CaptureWindow/Skill/SKILL.md:3]: Not adopting this change. The name uloop-screenshot is intentional — it provides better discoverability for AI agents who would naturally search for 'screenshot' rather than the internal tool name 'capture-window'. This comment is also outdated as the SKILL.md changes were reverted.

File: Packages/src/Editor/Config/ToolSettings.cs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+            string directory = Path.GetDirectoryName(SettingsFilePath);
+            if (!string.IsNullOrEmpty(directory) && !Directory.Exists(directory))
+            {
+                Directory.CreateDirectory(directory);
+            }
+
+            string json = JsonUtility.ToJson(settings, true);
+
+            Debug.Assert(json.Length <= McpConstants.MAX_SETTINGS_SIZE_BYTES,
+                "Settings JSON content exceeds size limit");
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Packages/src/Editor/Config/ToolSettings.cs:42]: Not adopting this change. This project uses Contract Programming with Debug.Assert for internal invariants per the project's Fail Fast policy. The same pattern is used in ULoopSettings.cs. In Unity Editor (always debug builds), Debug.Assert halts execution on violation. These size limits are preconditions for internal code, not external input validation.

File: Packages/src/Editor/Config/ToolSettings.cs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            AtomicFileWriter.CleanupBackup(SettingsFilePath + ".bak");
+        }
+
+        public static bool IsToolEnabled(string toolName)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(toolName), "toolName must not be null or empty");
+
+            string[] disabledTools = GetSettings().disabledTools;
+            return !disabledTools.Contains(toolName);
+        }
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Packages/src/Editor/Config/ToolSettings.cs:56]: Fixed in 65f6683. Added null check after JsonUtility.FromJson in LoadSettings — if the deserialized object or its disabledTools array is null (e.g., from hand-edited JSON with "disabledTools": null), it falls back to a default ToolSettingsData instance.

File: Packages/src/Editor/Config/ToolSkillSynchronizer.cs [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+                CreateNoWindow = true
+            };
+            NodeEnvironmentResolver.SetupEnvironmentPath(startInfo, NodeEnvironmentResolver.FindNodePath());
+
+            await Task.Run(() =>
+            {
+                Process process = Process.Start(startInfo);
+                process?.WaitForExit();
+                process?.Dispose();
+            });
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Packages/src/Editor/Config/ToolSkillSynchronizer.cs:146]: Fixed in 65f6683. Updated RunSkillsInstall to: (1) use 'using' pattern for Process disposal, (2) drain stdout/stderr to prevent buffer deadlock, (3) log Debug.LogWarning for null process or non-zero exit code. Chose LogWarning over throwing since this is a background operation that shouldn't crash the Editor window.

File: Packages/src/Cli~/src/tool-settings-loader.ts [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  if (!content.trim()) {
+    return [];
+  }
+
+  let parsed: unknown;
+  try {
+    parsed = JSON.parse(content);
+  } catch {
+    return [];
+  }
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Packages/src/Cli~/src/tool-settings-loader.ts:46]: Not adopting this change. The settings file is written by our own AtomicFileWriter.Write() which uses File.WriteAllText with .NET default UTF-8 encoding (no BOM). Since we control both the writer and reader, BOM stripping is unnecessary. If a user hand-edits the file with a BOM-producing editor, JSON.parse fails gracefully via the existing try-catch and returns [] (all tools enabled), which is a safe fallback.

File: Packages/src/Editor/Config/ToolSettings.cs [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+                string json = File.ReadAllText(SettingsFilePath);
+
+                if (string.IsNullOrWhiteSpace(json))
+                {
+                    _cachedSettings = new ToolSettingsData();
+                    return;
+                }
+
+                ToolSettingsData loaded = JsonUtility.FromJson<ToolSettingsData>(json);
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Packages/src/Editor/Config/ToolSettings.cs:116]: Not adopting this suggestion. The null check here is input validation for user-editable external data (.uloop/settings.tools.json), not an internal invariant. Per the project's contract programming guidelines, Debug.Assert is for programmer mistakes, while external input uses validation. Defaulting to 'all tools enabled' when the file is corrupted is the intentionally safe fallback — it never silently disables tools the user expects to work.

File: Packages/src/Editor/Config/ToolSkillSynchronizer.cs [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+            await Task.Run(() =>
+            {
+                using Process process = Process.Start(startInfo);
+                if (process == null)
+                {
+                    Debug.LogWarning($"[uLoopMCP] Failed to start uloop process for: skills install {targetFlag}");
+                    return;
+                }
+
+                // Drain redirected streams to prevent buffer deadlock
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Packages/src/Editor/Config/ToolSkillSynchronizer.cs]: Fixed in 8a7978f. Changed to read stderr via ReadToEndAsync() while reading stdout synchronously, preventing potential buffer deadlock when both streams are redirected.
